eth_defi/aave_v3/events.py

```diff
@@ -0,0 +1,311 @@
+):
+    """Fetch all tracked Aave v3 events to CSV files for notebook analysis.
+
+    Creates couple of CSV files with the event data:
+
+    - `/tmp/aave-v3-{aave_network_name.lower()}-reservedataupdated.csv`
+
+    A progress bar and estimation on the completion is rendered for console / Jupyter notebook using `tqdm`.
+
+    The scan be resumed using `state` storage to retrieve the last scanned block number from the previous round.
+    However, the mechanism here is no perfect and only good for notebook use - for advanced
+    persistent usage like database backed scans, please write your own scan loop using proper transaction management.
+
+    .. note ::
+
+        Any Ethereum address is lowercased in the resulting dataset and is not checksummed.
+
+    :param json_rpc_url: JSON-RPC URL
+    :param start_block: First block to process (inclusive), default is block xxx (when Aave v3 xxx was created on mainnet)
+    :param end_block: Last block to process (inclusive), default is block xxx (1000 block after default start block)
+    :param aave_network_name: Network name, e.g. 'Polygon'
+    :param state: Store the current scan state, so we can resume
+    :param output_folder: Folder to contain output CSV files, default is /tmp folder
+    :param max_workers:
+        How many threads to allocate for JSON-RPC IO.
+        You can increase your EVM node output a bit by making a lot of parallel requests,
+        until you exhaust your nodes IO capacity. Experiement with different values
+        and see how your node performs.
+    :param log_info: Which function to use to output info messages about the progress
+    """
+    token_cache = TokenCache()
+    http_adapter = HTTPAdapter(pool_connections=max_workers, pool_maxsize=max_workers)
+    web3_factory = TunedWeb3Factory(json_rpc_url, http_adapter)
+    web3 = web3_factory(token_cache)
+    executor = create_thread_pool_executor(web3_factory, token_cache, max_workers=max_workers)
+    event_mapping = get_event_mapping(web3)
+    contract_events = [event_data["contract_event"] for event_data in event_mapping.values()]
+
+    # Start scanning
+    restored, restored_start_block = state.restore_state(start_block)
+    original_block_range = end_block - start_block
+
+    if restored:
+        log_info(f"Restored previous scan state, data until block {restored_start_block:,}, we are skipping {restored_start_block - start_block:,} blocks out of {original_block_range:,} total")
+    else:
+        log_info(
+            f"No previous scan done, starting fresh from block {start_block:,}, total {original_block_range:,} blocks",
+        )
+
+    # Prepare local buffers and files.
+    # Buffers is a context dictionary that is passed around
+    # by the event scanner.
+    buffers = {}
+
+    for event_name, mapping in event_mapping.items():
+        file_path = f"{output_folder}/aave-v3-{aave_network_name.lower()}-{event_name.lower()}.csv"
+        exists_already = Path(file_path).exists()
```

Comment:
Sanity check - pathlib.Pah already properly handles forward slashes as path separators in the raw path, right? Should not cause problems if someone runs this code on Windows?